finesse/hardware/st10_controller.py

```diff
@@ -0,0 +1,343 @@
+"""Code for interfacing with the ST10-Q-NN stepper motor controller.
+
+Applied Motions have their own bespoke programming language ("Q") for interfacing with
+their devices, of which we're only using a small portion here.
+
+The specification is available online:
+    https://appliedmotion.s3.amazonaws.com/Host-Command-Reference_920-0002W_0.pdf
+"""
+
+import logging
+from typing import Any, Optional
+
+from serial import Serial, SerialException, SerialTimeoutException
+
+from .stepper_motor_base import StepperMotorBase
+
+
+class ST10ControllerError(SerialException):
+    """Indicates that an error has occurred with the ST10 controller."""
+
+
+class ST10Controller(StepperMotorBase):
+    """An interface for the ST10-Q-NN stepper motor controller.
+
+    This class allows for moving the mirror to arbitrary positions and retrieving its
+    current position.
+    """
+
+    STEPS_PER_ROTATION = 50800
+    """The total number of steps in one full rotation of the mirror."""
+
+    def __init__(self, serial: Serial) -> None:
+        """Create a new ST10Controller.
+
+        Args:
+            serial: The serial device to communicate with the ST10 controller
+
+        Raises:
+            SerialException: Error communicating with device
+            SerialTimeoutException: Timed out waiting for response from device
+            ST10ControllerError: Malformed message received from device
+        """
+        self.serial = serial
+
+        # Check that we are connecting to an ST10
+        self._check_device_id()
+
+        # In case the motor is still moving, stop it now
+        self.stop_moving()
+
+        # Move mirror to home position
+        self._home_and_reset()
+
+        super().__init__()
```

Comment:
I would call this the first thing, so any customisations of this class are done on top of whatever the base class has setup.

Reply:
My only worry with doing that in this case is that the base class does a `pub.subscribe()` in its `__init__()` and ideally I think we want that to be the last piece of setup we do, partly so we can check that its the right device with `_check_device_id()` before proceeding but also so that there is no chance of the motor being moved in response to a `stepper.move` message before the commands to move it to the home position have been sent (otherwise it'll move to the wrong position). The message would have to be sent from another thread for this to happen, so I don't think it'll happen with the way the code is written now, but we might as well make it thread-safe if we can!